step-55/step-55.cc

```diff
@@ -143,37 +143,50 @@ namespace Step55
 
     // The class A template class for a simple block diagonal preconditioner
     // for 2x2 matrices.
-    template <class PreconditionerA, class PreconditionerS>
+    template <class PreconditionerA, class PreconditionerS, typename VectorType>
     class BlockDiagonalPreconditioner : public Subscriptor
     {
     public:
       BlockDiagonalPreconditioner(const PreconditionerA &preconditioner_A,
-                                  const PreconditionerS &preconditioner_S);
+                                  const PreconditionerS &preconditioner_S,
+                                  const Nullspace<VectorType> &nullspace);
 
       void vmult(LA::MPI::BlockVector &      dst,
                  const LA::MPI::BlockVector &src) const;
 
     private:
       const PreconditionerA &preconditioner_A;
       const PreconditionerS &preconditioner_S;
+      const Nullspace<VectorType> &nullspace;
+
     };
 
-    template <class PreconditionerA, class PreconditionerS>
-    BlockDiagonalPreconditioner<PreconditionerA, PreconditionerS>::
+    template <class PreconditionerA, class PreconditionerS, typename VectorType>
+    BlockDiagonalPreconditioner<PreconditionerA, PreconditionerS, VectorType>::
       BlockDiagonalPreconditioner(const PreconditionerA &preconditioner_A,
-                                  const PreconditionerS &preconditioner_S)
+                                  const PreconditionerS &preconditioner_S,
+                                  const Nullspace<VectorType> &nullspace)
       : preconditioner_A(preconditioner_A)
       , preconditioner_S(preconditioner_S)
+      , nullspace(nullspace)
     {}
 
 
-    template <class PreconditionerA, class PreconditionerS>
-    void BlockDiagonalPreconditioner<PreconditionerA, PreconditionerS>::vmult(
+    template <class PreconditionerA, class PreconditionerS, typename VectorType>
+    void BlockDiagonalPreconditioner<PreconditionerA, PreconditionerS, VectorType>::vmult(
       LA::MPI::BlockVector &      dst,
       const LA::MPI::BlockVector &src) const
     {
       preconditioner_A.vmult(dst.block(0), src.block(0));
       preconditioner_S.vmult(dst.block(1), src.block(1));
+
+      // Projection for block(0).
+      double inner_product1 = nullspace.basis[0].block(0)*dst.block(0);
+      dst.block(0).add( -1.0*inner_product1, nullspace.basis[0].block(0));
```

Comment:
You can do the projcetion for both at once:
```suggestion
      double inner_product1 = nullspace.basis[0*dst;
      dst.add( -1.0*inner_product1, nullspace.basis[0]);
```